### ISSUE TITLE
Launching v5 Beta

### ISSUE THREAD

**Issue by fcusson** [fondberg/spotcast#478]
> [!NOTE]
> I am launching the open beta for spotcast `v5.0.0-b0`. If you are interested in testing the newest build. Starting today, you will have access to pre-release builds for `v5` .

This one was something to build. This here is the first pre-release for Spotcast Version 5. This new version of Spotcast is a complete rewrite of the application to modernise the project to current expectation from Home Assistant and also remove our dependency on the main Spotify Integration as much as possible. There are still a few items to solve, but I feel the code is at a place where testing by a larger audience will be a benefit. For anyone wanting to test out this new version.

# Roadmap

- [x] Rewrite of the core system with its own OAuth manager
- [x] Separation of spotcast functionalities is separate services
- [x] Rebuild of the websocket API
- [x] Creation of spotify connect device media_players
- [x] Creation of Service Sensors
- [ ] Creating our own media library browser
- [ ] Creation of a dedicated podcast service
- [x] Creation of a yaml config import
- [x] Creation of a configuration flow to enable configuring Spotcast from the UI
- [x] Creation of an option flow to permit modifying options without restarting Home Assistant
- [x] Creation of a reauth process for when credentials expires
- [x] Provide a comprehensive UI text definition
- [x] Provide translations for the different Spotcast Interfaces
- [x] Creation of a logo and icon for Spotcast
- [x] Register the icons and logos in Home Assistant branding

# Configuration

The configuration for Spotcast has changed drastically in this build. YAML configuration is no longer available and has been replaced with a UI configuration. Expect the following:

1. As there is no YAML import at the moment, you will have to remove your entry for spotcast in the `configuration.yaml`
2. A Spotify Developper Application is now required to manage the OAuth token. If you have the official Spotify Integration, you can use the same.
3. All documentation about the new version of spotcast can be found in the [dev branch](https://github.com/fondberg/spotcast/tree/dev)
4. Expect frequent change in the coming weeks. 

# Installing pre-release version in HACS

HACS will not offer pre-release version of integrations unless you specify you want to install them. To install V5:

1. Go to the following link

[![Open your Home Assistant instance and open a repository inside the Home Assistant Community Store.](https://my.home-assistant.io/badges/hacs_repository.svg)](https://my.home-assistant.io/redirect/hacs_repository/?owner=fondberg&repository=spotcast)

2. Click the three dots in the upper right corner and select `Redownload`

![Image](https://github.com/user-attachments/assets/5a706a3d-4d98-4a82-85c1-818d32d37680)

3. In the menu, select to most recent `pre-release` version

![Image](https://github.com/user-attachments/assets/dbca95b1-a765-4d1a-b6df-506cb2ae7c32)

Thank you in advance for anyone wanting to test out this pre-release. I have been stress testing it for a bit now and have had very good result, especially in terms of performance for starting playback. I'll be waiting for bugs report, as I am sure there are a lot of edge cases that I did not anticipate

**Comment by wildenrou** [comment]
Spotcast was already my favorite and most useful integration among the 58 integrstions and add-ons I use on my HASS instance. Now testing the v5 and I have to say, this is truly amazing and has huge potential for other use cases. Thanks again for all this great work and the last quick upgrades. You're on fire!!!

**Comment by BeastleeUK** [comment]
I have a very niche requirement for this as part of a custom alarm clock blueprint I designed. For some reason updating to HA 2024.11.x broke Spotcast so I figured I'd use the beta and thankfully it works. V5.x seems to be a fantastic evolution so far and my automation seems to be back to working too! Keep up the great work!

The only thing I've come stuck on is that I use a select helper to determine the account to use, this had a template that set the account dynamically but with the move to the integration ID being used I have no idea how to retrieve this as part of a template, can anyone advise how I do this?

**Comment by fcusson** [comment]
Hi @BeastleeUK,

The entry id was supposed to be part of the profile sensor. I'll add it in the next build. Sorry my mistake. You should be able with that to use templating to fetch the proper attributes based on the profile sensor used.

**Comment by mikevanes** [comment]
@fcusson I am writing a new version of the spotify card at the moment for the V5 integration. It seems to be that the `spotcast/castdevicesspotcast/castdevices` doesn't return anything at the moment. If I am not mistaken this does not require any account to be passed to the socket right?

**Comment by fcusson** [comment]
@mikevanes castdevices is not looking into an account, it is simply calling the cast integration and returning results from that domain.

```python
@websocket_wrapper
async def async_get_cast_devices(
    hass: HomeAssistant,
    connection: ActiveConnection,
    msg: dict
):
    """Gets a list playlists from an account

    Args:
        - hass(HomeAssistant): the Home Assistant Instance
        - connection(ActiveConnection): the Active Websocket connection
            object
        - msg(dict): the message received through the websocket API
    """

    entities: dict[str, CastDevice] = await async_entities_from_integration(
        hass,
        "cast",
        ["media_player"],
    )

    devices = []

    for id, entity in entities.items():

        cast_info = entity._cast_info.cast_info

        devices.append({
            "entity_id": id,
            "uuid": str(cast_info.uuid),
            "model_name": cast_info.model_name,
            "friendly_name": cast_info.friendly_name,
            "manufacturer": cast_info.manufacturer,
        })

    connection.send_result(
        msg["id"],
        {
            "total": len(devices),
            "devices": devices
        },
    )
```

New replies should look as documented [here](https://github.com/fondberg/spotcast/blob/dev/docs/websocket/cast_devices.md). I tested  and can confirm I am receiving the expected reply from the websocket with this call:

````json
{
    "id": 5,
    "type": "spotcast/castdevices"
}
````

**Comment by fcusson** [comment]
@mikevanes are you trying to manipulate the reply in anyway? I harmonized all the responses format, so that might be why its not working as expected

**Comment by mikevanes** [comment]
No I am not casting the reply in any way. The reply I am getting is:

`{
  "total": 0,
  "devices": []
}`

**Comment by fcusson** [comment]
hmmm, very weird, and you have cast devices available in Home Assistant?

The only way to get that reply without error from the websocket handler would be for `async_entities_from_integration` to provide an empty dictionary, which would mean HomeAssistant couldn't provide any `media_players` that are part of the `cast` domain. Lets move that to a separate ticket, I'll want that to be a specific point of discussion. You can open it under re-enabling spotify-card, we'll use that thread for any issue with what you need for your component

**Comment by fcusson** [comment]
@DwainTR. Didn't want to continue the conversation on the spotify integration ticket.

Just wanted to clarify, Spotcast is a custom integration not an Add-On, if you have HACS, you can simply search for it and install.

**Comment by BeastleeUK** [comment]
> Hi @BeastleeUK,
> 
> The entry id was supposed to be part of the profile sensor. I'll add it in the next build. Sorry my mistake. You should be able with that to use templating to fetch the proper attributes based on the profile sensor used.

Thanks for this, I can now find the Id, however I still can't use my automation as the only sensors spotcast provides have the name of the account in them so can't be dynamically selected. Is it possible to have a sensor that lists all the accounts as objects, i.e. spotcast.accounts with each account having the name and id as attributes? Happy to put this as a separate FR if you think it's feasible.

Basically I want to make the account selectable via an input select helper, this then gets looked at by the the alarm clock routine and plays the music of the selected account on the device, selected from a different helper. The plan will be to have a n automation that dynamically updates the input select helper with the data from the account list so that it's always in sync. This will allow me to manage multiple alarms in the household. The alarm fades in Spotify and Hue lighting for a gradual wake up.

**Comment by fcusson** [comment]
@BeastleeUK 

> Basically I want to make the account selectable via an input select helper, this then gets looked at by the the alarm clock routine and plays the music of the selected account on the device,

Didn't implement a global sensor because for something like that I use jinja templates. You could use something of the kind for your usecase:

````jinja
{% set name_to_id = {} %}
{% for entity in states.sensor if entity.entity_id.startswith('sensor.spotcast_') and entity.entity_id.endswith('spotify_profile') %}
  {% if entity.attributes.display_name == states('input_select.spotcast_user') %}
    {{ entity.attributes.entry_id }}
  {% endif %}
{% endfor %}
````

That could either be a template sensor that listen to the spotcast_user input select or directly in your automation. I would prefer the first option, because then its easy to use elsewhere.

If you still believe adding it to the code base would be important, make a PR, but your sensor will have to be global, so ensure you are not rebuilding it on a second account setup.

**Comment by BeastleeUK** [comment]
> @BeastleeUK
> 
> > Basically I want to make the account selectable via an input select helper, this then gets looked at by the the alarm clock routine and plays the music of the selected account on the device,
> 
> Didn't implement a global sensor because for something like that I use jinja templates. You could use something of the kind for your usecase:
> 
> **edit: found a much cleaner way with label_entities**
> 
> ```
> {% for entity in label_entities('spotcast_account') %}
>   {% if state_attr(entity, 'display_name') == states('input_select.spotcast_user') %}
>     {{ state_attr(entity, 'entry_id') }}
>   {% endif %}
> {% endfor %}
> ```
> 
> That could either be a template sensor that listen to the spotcast_user input select or directly in your automation. I would prefer the first option, because then its easy to use elsewhere.
> 
> If you still believe adding it to the code base would be important, make a PR, but your sensor will have to be global, so ensure you are not rebuilding it on a second account setup.

Sorry for the slow response, I ran out of time on the day and only just back to laptop.

Thank you for the script, and the original one you posted. I'm reviewing how the label_entities works, then I can put this into place on the alarm automation itself. I've used the first script you posted to enable me to update the options in the input_select, I've had to run it manually as there's no trigger option for 'new entity added' but it works a treat!

**Comment by BeastleeUK** [comment]
I've got that working with the above code and applying the label to the profile entity, thanks so much. Next job will be to make it a template sensor and just reference that in the automation.

When testing this change it reported 'Error: Account `xxxxxxxxx` has no known or active playback. Nothing to transfer' even though it normally just picks up where I left off. I hit play and pause on my phone then it transferred fine but I've not had to do this previously. Hopefully it will work in the morning and this was just a Spotify glitch.

**Comment by fcusson** [comment]
@BeastleeUK, Spotify seem to have changed the way they report long lasting playback now. They provide an empty reply as soon as you close the app that was playing. It was possible on the past to restart the playback long after the app was closed. 

I've implemented a way to restart playback from a long gone playback state in the latest build, that should solve your issue. Currently it only works for album or playlist but I'm working on implementing a solution for artists and podcasts

**Comment by BeastleeUK** [comment]
Weird that it was fine Tuesday but not last night/this morning. I rarely play anything other than my liked songs list on repeat with suggestions. I'll see what I can do with the latest build to make it work for this setup.

Thanks again.

**Comment by gmcmicken** [comment]
Can we have support for "recently played" playlist on the websocket api? A while back I added this functionality into the previous version of spotcast but it's been dropped it seems. [https://github.com/fondberg/spotcast/pull/257](https://github.com/fondberg/spotcast/pull/257)

**Comment by mikevanes** [comment]
@gmcmicken using the `view` websocket endpoint you can retrieve the recently played playlists. It should have documentation on how to do so.

**Comment by gmcmicken** [comment]
> [@gmcmicken](https://github.com/gmcmicken) using the `view` websocket endpoint you can retrieve the recently played playlists. It should have documentation on how to do so.

Wonderful, my apologies

**Comment by litecross91** [comment]
When I start a podcast with the script below, only the first episode plays and then it stops. This is what version 4.x did. Is there a workaround to keep it going to the next episode? Thanks!

```
- action: spotcast.play_media
    data:
      media_player:
        entity_id: media_player.kitchen
      spotify_uri: spotify:show:xxxxxxxxxxxxxxxxxxxxxxx
```

**Comment by litecross91** [comment]
I opened: https://github.com/fondberg/spotcast/issues/516
Thank you!

**Comment by XalaTheShepard** [comment]
@fcusson  : in V4 there was an option by adding 'random_song: true' to the data to shuffle the music. Is this also available in V5? If yes, what should be added? 
Except for the shuffle modus, everything is working perfectly. 

Many thanks in advance and for all the support. 

Kind regards, 

Robin

**Comment by fcusson** [comment]
@XalaTheShepard that's part of the data option now. See [documentation](https://github.com/fondberg/spotcast/blob/dev/docs/services/play_media.md#data-dict).

**Comment by Mincka** [comment]
Hello, the integration is working great for me. Thanks. 😃 

Quick question: How to update `sp_dc` and `sp_key` in the v5 integration when they expire?
I don't see them in "Configure".

**Comment by XalaTheShepard** [comment]
> [@XalaTheShepard](https://github.com/XalaTheShepard) that's part of the data option now. See [documentation](https://github.com/fondberg/spotcast/blob/dev/docs/services/play_media.md#data-dict).

@fcusson : many thanks. Everything is working like before again. :)

**Comment by Stefterv** [comment]
Hi! The integration is working great, thanks! One feature that I was using of V4 was the ability to transfer playback to my Sonos devices through the transfer id. The new transfer playback action limits to devices it has discovered, which sadly my Sonos devices are not discovered.

**Comment by fcusson** [comment]
@Stefterv for devices to be transferable it needs to be registered to the account and part of the available list from the API. Otherwise there is a registration blob that must be created and sent to register the device. 

As we were never creating that handshake in the past with unregistered devices, it shouldn't be working if the device was not preregistered and if it was preregistered it should show up in your spotcast media players

**Comment by Stefterv** [comment]
That's the fun part with Sonos devices, they are not visible in the API but are visible in the Spotify app and website [more details](https://www.reddit.com/r/homeassistant/comments/13ya97g/sonos_and_spotify_interaction_issues/) It seems like they use an older (private) API

**Comment by fcusson** [comment]
@Stefterv no that's [zeroconf](https://developer.spotify.com/documentation/commercial-hardware/implementation/guides/zeroconf) that just mean they were discovered on your network.
 
A network device is not registered to an account. Otherwise a only that account in the network could access them if that was the case. So there is a zeroconf setup to find over network the device and then register it using a registration blob. That blob is undocumented by spotify as they want to have control over its way of working.

People have reversed engineered the blob for user/password but access token blob is not fully working yet. I'm working on an implementation, but its not ready.

**Comment by fcusson** [comment]
@Stefterv yes spotifyplus would be an option. They are using the user/password blob login at the moment. I decided to go against using user/password in my case. Might take a bit more time, but we should have a true zeroconf token based registration down the line in the end, which would be quite sueful since Spotify is talking about removing user/password blobs as an option eventually

**Comment by Stefterv** [comment]
@fcusson Interesting, I have Spotify Plus configured without a Username/Pass at the moment and it works with transferring to the Sonos devices. I agree with you that I don't want to configure my login credentials anywhere

**Comment by fcusson** [comment]
@Stefterv is that with Sonos in their standard custom mode or spotify connect mode?

Also, I wouldnt be personally capable of working on this implementation directly as I do not own any Sonos devices. Someone with the devices will have to integrate into the project

The good news is that V5 makes it much easier to integrate new devices especially if the media player is part of home assistant all we need is a way of registering the account to the device and to provide the device id to the api

**Comment by fcusson** [comment]
@paull27 if you want to contribute on Sonos development their is already a ticket for that #376.

**Comment by geoffoxholm** [comment]
I'm curious about the creation and removal of media_player entities. I see about 62 media players created (including re-creations of the same id) each day. I think these are the "spotify connect" media_player entities you referred to. I haven't determined a use for them, but I'm also not sure how somebody could with them being added and removed. I don't exactly have a problem with this, but I wanted to make sure it was intentional. On my end I have an automation that makes a HA notification when entities are added (because I like to know when an integration adds new functionality, or if a noisy entity is added that I may want to disable). So this behavior from spotcast is a bit distracting for me.

**Comment by fcusson** [comment]
@geoffoxholm could you provide some details about the kind of devices being created? Some are very stable like phone and desktop app, but some others less so.

There is definitely a use for them like starting playback on a desktop or phone through automations.

Devices are created based on Spotify device Id. If you have a specific type of devices that doesn't retain a stable device Id, that might be the issue.

Also I saw that spotify started to register jam sessions as "available devices", this is something I'm looking into.

**Comment by Mincka** [comment]
> Quick question: How to update `sp_dc` and `sp_key` in the v5 integration when they expire? I don't see them in "Configure".

Up 😇

**Comment by fcusson** [comment]
> > Quick question: How to update `sp_dc` and `sp_key` in the v5 integration when they expire? I don't see them in "Configure".
> 
> Up 😇

You will be asked to reconfigure when this happens

**Comment by geoffoxholm** [comment]
> [@geoffoxholm](https://github.com/geoffoxholm) could you provide some details about the kind of devices being created? Some are very stable like phone and desktop app, but some others less so.

Looking through it, almost all of them are:

```
 media_player.web_player_chrome_{account}_spotcast
```

**Comment by fcusson** [comment]
@geoffoxholm then that's normal, web player are ephemeral the spotify device id is not stable. They should be created and removed automatically. This is by design.

You could turn off notifications from spotcast for web player in your automation. That should be trivial.

**Comment by pezlin** [comment]
Sorry for stupid question but if I like to have four Spotify users using Spotcast should I set up for one OAuth for each user?

**Comment by fcusson** [comment]
@pezlin no stupid question here.

You do not need to create a new Oauth application per user, you can simply use you main one, but Spotify Dev application in development mode (which everyone uses here normally), you need to add the user to the user management tabs (25 user except app developper permitted).

![Image](https://github.com/user-attachments/assets/3b8918ee-bf87-4090-9ee9-3cdf0818cf81)

Once the email is in your user management list, the OAuth enrollment will work for them

**Comment by pezlin** [comment]
@fcusson I have the email addresses already in the User Management list but still gets below message when trying to one of these accounts to spotcast.

<img width="462" alt="Image" src="https://github.com/user-attachments/assets/be41b3a5-f0eb-45ef-8f69-3efe7ea692ac" />

**Comment by pezlin** [comment]
Is Spotify Card-V2 expected to work as usual with Spotcast Beta? For me it is only showing Loading...

**Comment by pezlin** [comment]
> [@pezlin](https://github.com/pezlin), not yet, [@mikevanes](https://github.com/mikevanes) has been working on a rework of the custom card

Received info from @mikevanes that the rework has been temporarily halted. Do you know if any other card is available or anyone else is working on such a card?

**Comment by pezlin** [comment]
@fcusson Thank you for your reply. I'm unfortunately no developer so can't assist in that task. Hopefully someone with the knowledge will be able to do so in the near future.

**Comment by notpurum** [comment]
I've faced a problem that I have (probably) entered some wrong credentials, when setting up the v5 Spotcast and receive "INVALID_CLIENT: Invalid client" error, when visiting the external site. The problem is, I now can not undo it and try entering the credentials again — I immediately get the "You must go to the external site" message after installing the v5 and get the error on the website.
What I have tried:
- returning to v4 and then again to v5
- deleting the integration
- removing the integration folder completely from the custom_components folder

Nothing helps

**Comment by fcusson** [comment]
#543

**Comment by ironcrafter54** [comment]
Btw this is already probably common knowledge but the spotifyplus developer has found a solution to this problem. https://github.com/thlucas1/homeassistantcomponent_spotifyplus/wiki/Device-Configuration-Options#spotify-desktop-player-authentication-configuration

**Comment by fcusson** [comment]
@BeastleeUK yes I'm aware, I'm talking with thlucas about potential fixes in the ticket I referenced.

**Comment by fcusson** [comment]
As we are moving to the alpha for `V6` I will be closing this thread.
